FractionalWrapper

davidbrai opened the pull request:
FractionalWrapper: a thin wrapper over a Yearn Vault
YearnVaultMock: a mock of a Yearn Vault with configurable price-per-share

As specified in: https://github.com/yieldprotocol/mentorship2022/issues/4

Deployed to optimism kovan:
FunToken: https://kovan-optimistic.etherscan.io/address/0xe63ccea54d1c53d0ab9444df26c69045ff64ad1d
YearnVaultMock: https://kovan-optimistic.etherscan.io/address/0x6e536eb178eb62586ae8a1a42133e343b22a4225
FractionalWrapper: https://kovan-optimistic.etherscan.io/address/0x9db9f594b9596b1751f29e3bc161b4b0b36d4ba3

davidbrai committed:
feat: add FractionalWrapper and YearnVaultMock

alcueca commented on src/assignment-4/FractionalWrapper.sol:
```diff
@@ -0,0 +1,75 @@
+        uint256 numShares = yvToken.deposit(amount);
+
+        // Wrapper sends Wrapper shares to user
+        _mint(msg.sender, numShares);
```
checks-effects-interactions, the `_mint` should happen before the external call.

davidbrai replied:
@alcueca I don't think I can do the `_mint` before the external calls.

* for the `_mint(...)` call, I need to know the `numShares`, which is returned from `yvToken.deposit(...)`, so I must call it first.
* and in order to call `yvToken.deposit(...)`, I need to first transfer the underlying token into the wrapper contract, otherwise it may not have enough balance

alcueca replied:
You are right, actually. I had never before come across an instance where checks-effects-interactions couldn't be implemented. Maybe that's because I haven't worked much on yield aggregators.

In this case, since CEI cannot be implemented, we can carefully check if there are reentrancy risks (in this case the function ordering seems to make it safe), and we could look into adding a [reentrancy blocker](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/ReentrancyGuard.sol) if we are not 100% sure.

alcueca commented on src/assignment-4/FractionalWrapper.sol:
```diff
@@ -0,0 +1,75 @@
+    ///     3. Yearn Vault sends `numShares` of `yvToken` to wrapper
+    ///     4. Wrapper sends `numShares` of wrapper shares to user
+    /// @param amount The amount of underlying token to deposit
+    function deposit(uint amount) public {
```
It's good practice to always return any values that are obtained during execution, in this case `numShares`. It helps enormously to integrate this contract with others.

davidbrai replied:
makes sense. fixed here: https://github.com/davidbrai/yield-mentorship/pull/5/commits/f67495b08eca3daa7a8f576a1386d694cf707e3c

alcueca commented on src/assignment-4/FractionalWrapper.sol:
```diff
@@ -0,0 +1,75 @@
+    ///     3. Yearn Vault sends `amount` of `token` to wrapper
+    ///     4. Wrapper sends `amount` of `token` back to user
+    /// @param numShares The number of wrapper shares to burn
+    function burn(uint numShares) public {
```
In this case, we would want to return `amount`.

davidbrai replied:
:+1: fixed here: https://github.com/davidbrai/yield-mentorship/pull/5/commits/f67495b08eca3daa7a8f576a1386d694cf707e3c

alcueca commented on src/assignment-4/FractionalWrapper.sol:
```diff
@@ -0,0 +1,75 @@
+        _burn(msg.sender, numShares);
+
+        // Withdraw from YVault, get back underlying token
+        uint256 amount = yvToken.withdraw(numShares);
```
Probably you want to use a more consistent variable naming. If you use `token` and `yvToken`, then maybe use `tokenAmount` and `yvTokenAmount` when you deal with token and yvToken amounts :D

davidbrai replied:
👍 fixed here: https://github.com/davidbrai/yield-mentorship/pull/5/commits/f67495b08eca3daa7a8f576a1386d694cf707e3c

alcueca commented on src/assignment-4/YearnVaultMock.sol:
```diff
@@ -0,0 +1,67 @@
+        token = IERC20(token_);
+    }
+
+    function pricePerShare() external view returns (uint256) {
```
So, the mocks don't get natspec love?

davidbrai replied:
:+1: fixed: https://github.com/davidbrai/yield-mentorship/pull/5/commits/93c82b91814228389e267b81dcc921fd9c5e5ffe

alcueca reviewed (commented):
Some minor points, nothing serious. I still have to review the tests.

davidbrai committed:
fix code review comments

1. return values from deposit and withdraw functions
2. rename variables for more consistency

davidbrai committed:
fix: oops

davidbrai committed:
add natspec to YearnVaultMock

alcueca commented on src/assignment-4/FractionalWrapper.sol:
```diff
@@ -0,0 +1,81 @@
+    ///     4. Wrapper sends `amount` of `token` back to user
+    /// @param wrapperTokenAmount The number of wrapper shares to burn
+    /// @return The amount of `token` tokens sent to the user
+    function burn(uint wrapperTokenAmount) public returns (uint256) {
```
I would call this one `withdraw` to match the underlying yvToken, or I would rename `deposit` into `mint` otherwise to match the ERC20 convention.

davidbrai replied:
👍  fixed: https://github.com/davidbrai/yield-mentorship/pull/5/commits/8c5491ed9872de6bc90b69dc01ad7b0758abdcae

alcueca commented on src/assignment-4/FractionalWrapper.sol:
```diff
@@ -0,0 +1,81 @@
+        return yvTokenAmount;
+    }
+
+    /// @notice Burns a specified amount of wrapper shares and returns the underlying token to the user
```
I've got a stretch goal for you, please implement a function in FractionalWrapper.sol to recover an exact amount of `token`. Please do not modify YearnVaultMock.sol to achieve this.

```
function withdrawExact(uint256 tokenAmount) public returns (uint256 wrapperAmount);
```

davidbrai replied:
done: https://github.com/davidbrai/yield-mentorship/pull/5/commits/e47e6d45754382bb326e1545a53cf99322f47aeb

alcueca commented on test/assignment-4/FractionalWrapper.t.sol:
```diff
@@ -0,0 +1,128 @@
+    function setUp() public virtual {
+        token = new FunToken();
+        yvToken = new YearnVaultMock(address(token), "Yearn Vault Mock", "YVMock", token.decimals());
+        yvToken.setPricePerShareMock(2 * RAY);
```
Please set a weirder price than this, something like `(18 * RAY) / 17`.

davidbrai replied:
@alcueca `testNumSharesWhenPriceHasFractions` is testing with a weirder price (12.345)
Do you feel like it's important all the tests have a weird price?

alcueca commented on test/assignment-4/FractionalWrapper.t.sol:
```diff
@@ -0,0 +1,128 @@
+}
+
+contract ZeroStateTest is ZeroState {
+    function testDepositingIntoWrapperSendsTokensToYVault() public {
```
You can merge the following 4 test cases into one. I do one test case per execution path, and that test includes all events and values.

davidbrai replied:
@alcueca I think this is more of a code style thing. I'm used to writing unit-tests where each test tests for "one thing". For future tests, I can write them so that they include all the things happening in the executing path

alcueca reviewed (changes requested):
Your code is so good, I gave you a fun stretch goal.

davidbrai committed:
refactor: rename method

davidbrai committed:
feat: add withdrawExact

davidbrai committed:
fix bug withdrawExact

division needs to be rounded up because multiplication in Yearn Vault is rounded down

davidbrai committed:
remove unused variable
